File: medium/cpp/2962.cpp
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description/

class Solution
{
public:
    long long countSubarrays(std::vector<int>& nums, int k)
    {
        const auto val = *std::max_element(std::cbegin(nums), std::cend(nums));
        const auto n = std::size(nums);

        auto cnt = 0;
        auto res = 0LL;

        for (auto left = 0, right = 0; left < n; left++)
        {
            while (right < n && cnt < k)
            {
                if (nums[right] == val)
                {
                    cnt++;
                }

                right++;
            }

            if (cnt == k)
            {
                res += n - right + 1;
            }

            if (nums[left] == val)
            {
                cnt--;
            }

            if (cnt < k && right == n)
            {
                break;
            }
        }

        return res;
    }
};

File: medium/python/2962.py
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description/

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        val = max(nums)
        left, right = 0, 0
        cnt = 0
        res = 0

        while left < len(nums):
            while cnt < k and right < len(nums):
                if nums[right] == val:
                    cnt += 1
                right += 1

            if cnt < k:
                break

            res += len(nums) - right + 1

            if nums[left] == val:
                cnt -= 1

            left += 1

        return res
        